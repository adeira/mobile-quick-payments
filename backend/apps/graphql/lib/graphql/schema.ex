defmodule GraphQL.Schema do
  use Absinthe.Schema

  import_types(GraphQL.Scenes.AvailableScenes)

  # TODO: implement `node` query for Payment, Client and Retailer types

  query name: "RootQuery" do
    field(:scenes, :AvailableScenes) do
      # passthrough
      resolve(fn _, _ -> {:ok, %{}} end)
    end
  end
end
