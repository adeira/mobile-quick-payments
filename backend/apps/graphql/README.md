```
â–² ~ http POST 127.0.0.1:2048 query="{__typename}" -v
POST / HTTP/1.1
Accept: application/json, */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 25
Content-Type: application/json
Host: 127.0.0.1:4040
User-Agent: HTTPie/0.9.9

{
    "query": "{__typename}"
}

HTTP/1.1 200 OK
cache-control: max-age=0, private, must-revalidate
content-length: 39
content-type: application/json; charset=utf-8
date: Sat, 22 Sep 2018 21:58:16 GMT
server: Cowboy

{
    "data": {
        "__typename": "RootQueryType"
    }
}
```
