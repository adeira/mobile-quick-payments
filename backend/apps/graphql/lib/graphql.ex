defmodule GraphQL do
  @moduledoc """
  Documentation for GraphQL.
  """

  @doc """
  Hello world.

  ## Examples

      iex> GraphQL.hello()
      :world

  """
  def hello do
    :world
  end
end
