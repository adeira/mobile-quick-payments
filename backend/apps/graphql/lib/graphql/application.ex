defmodule GraphQL.Application do
  # See https://hexdocs.pm/elixir/Application.html
  # for more information on OTP Applications
  @moduledoc false

  use Application

  def start(_type, _args) do
    # List all child processes to be supervised
    children = [
      # Starts a worker by calling: GraphQL.Worker.start_link(arg)
      # {GraphQL.Worker, arg},
      Plug.Adapters.Cowboy2.child_spec(
        scheme: :http,
        plug: GraphQL.Endpoint,
        options: [port: 2048]
      )
    ]

    # See https://hexdocs.pm/elixir/Supervisor.html
    # for other strategies and supported options
    opts = [strategy: :one_for_one, name: GraphQL.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
