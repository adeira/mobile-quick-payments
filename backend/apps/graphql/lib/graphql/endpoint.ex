defmodule GraphQL.Endpoint do
  use Plug.Router

  plug(:match)

  plug(Plug.Parsers,
    parsers: [:urlencoded, :multipart, :json],
    pass: ["*/*"],
    json_decoder: Poison
  )

  plug(:dispatch)

  get "/" do
    send_resp(conn, 200, "TODO: README (looking for /graphiql ?)")
  end

  post("/",
    to: Absinthe.Plug,
    init_opts: [
      schema: GraphQL.Schema,
      analyze_complexity: true,
      max_complexity: :infinity
    ]
  )

  get("/graphiql",
    to: Absinthe.Plug.GraphiQL,
    init_opts: [
      schema: GraphQL.Schema
    ]
  )

  match _ do
    send_resp(conn, 404, "oops 404")
  end
end
