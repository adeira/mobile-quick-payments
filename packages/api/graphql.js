// @flow

import fs from 'fs';
import path from 'path';
import { buildSchema } from 'graphql';

import { Payments as DatabasePayments } from './src/InMemoryDatabase';

export const schema = buildSchema(
  fs.readFileSync(path.join(__dirname, 'schema.graphql'), 'utf8'),
);

export const resolvers = {
  scenes: () => ({
    dashboard: {
      payments: ({ clientId }: {| clientId: number |}) => {
        return DatabasePayments.filter(
          payment => payment.clientId === clientId,
        );
      },
    },
  }),
};
