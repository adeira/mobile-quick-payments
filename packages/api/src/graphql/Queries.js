// @flow

import {GraphQLObjectType, GraphQLNonNull, GraphQLID} from 'graphql';

import {
  Clients as DatabaseClients,
  Payments as DatabasePayments,
  Retailers as DatabaseRetailers,
} from '../InMemoryDatabase';

module.exports = new GraphQLObjectType({
  name: 'Query',
  fields: {
    scenes: {
      type: require('../generated/GraphQLAllAvailableScenesType'),
      resolve: () => {
        // TODO: this should be replaced with generated schema - see src/__generated__/README.md
        return {
          dashboard: {
            // TODO: pagination
            payments: ({clientId}: {|+clientId: string|}) => {
              return DatabasePayments.filter(payment => payment.clientId === clientId).map(
                payment => ({
                  ...payment,
                  client: DatabaseClients.find(client => client.id === payment.clientId),
                  retailer: DatabaseRetailers.find(retailer => retailer.id === payment.retailerId),
                }),
              );
            },
          },
          payment: {
            // TODO: require client ID as well? (or authorize via headers)
            checkStatus: ({paymentId}: {|+paymentId: string|}) => {
              return DatabasePayments.find(payment => payment.id === paymentId);
            },
          },
        };
      },
    },

    node: {
      type: require('../generated/GraphQLNodeType'),
      args: {
        id: {
          type: GraphQLNonNull(GraphQLID),
        },
      },
      resolve: (_, args) => {
        // TODO: check the ID! and return appropriate object
        // (currently returns only Payment structure)
        return {
          id: args.id,
          total: {
            amount: 123,
            currency: 'MXN',
          },
        };
      },
    },
  },
});
