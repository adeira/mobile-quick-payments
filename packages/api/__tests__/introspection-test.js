// @flow

import {graphql} from 'graphql';

import {schema, resolvers} from '../graphql';

it('works', async () => {
  await expect(
    graphql(
      schema,
      `
        {
          __schema {
            directives {
              name
            }
            types {
              name
              fields {
                name
                type {
                  name
                }
                isDeprecated
              }
            }
          }
        }
      `,
      resolvers,
    ),
  ).resolves.toMatchSnapshot();
});
