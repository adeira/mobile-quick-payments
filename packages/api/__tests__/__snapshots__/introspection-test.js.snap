// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "data": Object {
    "__schema": Object {
      "directives": Array [
        Object {
          "name": "skip",
        },
        Object {
          "name": "include",
        },
        Object {
          "name": "deprecated",
        },
      ],
      "types": Array [
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "scenes",
              "type": Object {
                "name": "AllAvailableScenes",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "node",
              "type": Object {
                "name": "Node",
              },
            },
          ],
          "name": "Query",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "dashboard",
              "type": Object {
                "name": "DashboardScene",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "payment",
              "type": Object {
                "name": "PaymentScene",
              },
            },
          ],
          "name": "AllAvailableScenes",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "payments",
              "type": Object {
                "name": null,
              },
            },
          ],
          "name": "DashboardScene",
        },
        Object {
          "fields": null,
          "name": "ID",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "id",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "status",
              "type": Object {
                "name": "PaymentStatus",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "total",
              "type": Object {
                "name": "Money",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "client",
              "type": Object {
                "name": "Client",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "retailer",
              "type": Object {
                "name": "Retailer",
              },
            },
          ],
          "name": "Payment",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "id",
              "type": Object {
                "name": null,
              },
            },
          ],
          "name": "Node",
        },
        Object {
          "fields": null,
          "name": "PaymentStatus",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "amount",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "currency",
              "type": Object {
                "name": "SupportedCurrency",
              },
            },
          ],
          "name": "Money",
        },
        Object {
          "fields": null,
          "name": "String",
        },
        Object {
          "fields": null,
          "name": "SupportedCurrency",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "id",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": "String",
              },
            },
          ],
          "name": "Client",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "id",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": "String",
              },
            },
          ],
          "name": "Retailer",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "checkStatus",
              "type": Object {
                "name": "Payment",
              },
            },
          ],
          "name": "PaymentScene",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "types",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "queryType",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "mutationType",
              "type": Object {
                "name": "__Type",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "subscriptionType",
              "type": Object {
                "name": "__Type",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "directives",
              "type": Object {
                "name": null,
              },
            },
          ],
          "name": "__Schema",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "kind",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "description",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "fields",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "interfaces",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "possibleTypes",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "enumValues",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "inputFields",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "ofType",
              "type": Object {
                "name": "__Type",
              },
            },
          ],
          "name": "__Type",
        },
        Object {
          "fields": null,
          "name": "__TypeKind",
        },
        Object {
          "fields": null,
          "name": "Boolean",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "description",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "args",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "type",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "isDeprecated",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "deprecationReason",
              "type": Object {
                "name": "String",
              },
            },
          ],
          "name": "__Field",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "description",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "type",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "defaultValue",
              "type": Object {
                "name": "String",
              },
            },
          ],
          "name": "__InputValue",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "description",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "isDeprecated",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "deprecationReason",
              "type": Object {
                "name": "String",
              },
            },
          ],
          "name": "__EnumValue",
        },
        Object {
          "fields": Array [
            Object {
              "isDeprecated": false,
              "name": "name",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "description",
              "type": Object {
                "name": "String",
              },
            },
            Object {
              "isDeprecated": false,
              "name": "locations",
              "type": Object {
                "name": null,
              },
            },
            Object {
              "isDeprecated": false,
              "name": "args",
              "type": Object {
                "name": null,
              },
            },
          ],
          "name": "__Directive",
        },
        Object {
          "fields": null,
          "name": "__DirectiveLocation",
        },
      ],
    },
  },
}
`;
